### ISSUE TITLE
Design/implement formatted printing

### ISSUE THREAD

**Issue by jasone** [BranchTaken/Hemlock#111]
The `Format` API is powerful, but quite difficult to use correctly. Both `Format` and `Printf` are quite complicated to implement from a type safety perspective, and format string ergonomics are less than optimal. Furthermore, the APIs make effect tracking extremely unwieldy. Design and implement a formatted printing API that cleanly supports custom formatting of inputs, effects tracking, file/stream/string output modes, ideally without excessive cognitive load on the programmer using the facility. This facility may best be implemented with explicit syntax support for format strings, in which case it won't be feasible to use in the bootstrap `Basis` library. But if no special syntax is required, favor implementing this in the bootstrap `Basis` in order to ease the OCaml -> Hemlock rewrite step of bootstrapping.

### Tasks:
- [x] Design formatter API.
- [x] Design formatted string syntax.
- [x] Extend design documentation for interpolated string syntax.
  + [x] Desugared form as semantic meaning
  + [x] `\\%` protection
- [x] Implement the `Fmt` module.
- [x] Implement `String.Fmt`.
- [x] Implement `File.Fmt`.
- [x] Implement `to_string` and `fmt` for all atomic types supported by formatted strings.
  + [x] `bool`
  + [x] `Uns`
  + [x] `Sint`
  + [x] `Intnb`-based types
  + [x] `Intw`-based types
  + [x] `real`
  + [x] `codepoint`
  + [x] `string`
- [x] Implement `fmt` for container types.
  + [x] `list`
  + [x] `array`
  + [x] `option`
  + [x] `result`
  + [x] `Ordmap.t`
  + [x] `Ordset.t`
  + [x] `Map.t`
  + [x] `Q.t`
  + [x] `Deq.t`
  + [x] `Stream.t`
- [x] Implement `fmt` for various types which provide `pp`.
  + [x] `unit`
  + [x] `Cmp.t`
  + [x] `Sign.t`
  + [x] `Real.Dir.t`
  + [x] `Real.Class.t`
  + [x] `Real.Parts.t`
  + [x] `Bytes.Slice.t`
  + [x] `Text.Cursor.t`
  + [x] `Realer.t`
- [x] Add `fmt` to interfaces in `FormattableIntf`.
- [x] Rename existing `Format`-related `pp*` symbols to `xpp*`.
- [x] Rename `fmt` in `FormattableIntf` to `pp`.
- [x] Refactor tests to use `Fmt` rather than `Format`.
  + [x] #190
  + [x] #192
  + [x] #176
  + [x] #180
  + [x] #200
  + [x] #203
  + [x] #181
  + [x] #177
  + [x] #206
  + [x] #209
  + [x] #207
  + [x] #208
  + [x] #205
  + [x] #191
  + [x] #197
  + [x] #189
  + [x] #195
  + [x] #183
  + [x] #184
  + [x] #202
  + [x] #173
  + [x] #198
  + [x] #193
  + [x] #204
  + [x] #196
  + [x] #201
  + [x] #199
  + [x] #179
  + [x] #214
  + [x] #211
  + [x] #216
  + [x] #212
  + [x] #213
  + [x] #210
  + [x] #175
  + [x] #185
  + [x] #186
  + [x] #188
- [x] Remove `xpp` from interfaces in `FormattableIntf`.
- [x] Remove all `Format`-related `xpp*` pretty printer functionality.

**Comment by jasone** [comment]
There are two parts to this design:

1) Formatted I/O API (mandatory)
2) Syntax for formatted strings (optional)

# API

We need to support the possibility of effectless formatting, specifically so that we can use formmatters to create strings without causing side effects. In contrast, OCaml's [Printf.fprintf](https://www.ocaml.org/api/Printf.html#VALfprintf) `%a`/`%t` specifiers support formatted printing only via effectful functions. The API we want requires that state and behavior be encapsulated; first-class modules provide a clean solution. What follows is written in Hemlock so that the effects semantics are explicit, but the API can be stripped of effect typing and implemented in OCaml.
The `Fmt.Formatter` module type provides the state/behavior encapsulation.

```hemlock
# Fmt.hmi ----------------------------------------------------------------------
type Formatter: Formatter (e: effect) =
    type t
    val fmt: string -> t >e-> t
```

Any given type can implement a `fmt` function similar to the following:

```hemlock
type t

# val to_string: t -> string

# val fmt >e: t -> Fmt.Formatter e >e-> Fmt.Formatter e
let fmt t (formatter >e: Fmt.Formatter e) =
    formatter |> Fmt.Formatter.fmt (to_string t)
```

The key point to note is that `fmt` has a parametric effect that depends on the formatter, but `fmt` is otherwise effectless. Thus `fmt` can be used in conjunction with a string/file/whatever formatter without regard for formatter effects.

Following is a sketch of the APIs needed to support formatted strings and I/O. The syntax design proposed thereafter depends upon this API.

```hemlock
# Fmt.hmi ----------------------------------------------------------------------
type Formatter: Formatter (e: effect) =
    type t
    val fmt: string -> t >e-> t

type just =
| Left
| Center
| Right

type sign =
| Implicit
| Explicit
| Space

type base =
| Bin
| Oct
| Dec
| Hex

val pad_default: codepoint
val just_default: just
val sign_default: sign
val alt_default: bool
val zpad_default: bool
val width_default: uns
val precision_default: uns
val base_default: base

# Fmt.hm -----------------------------------------------------------------------
let pad_default = ' '
let just_default = Right
let sign_default = Implicit
let alt_default = false
let zpad_default = false
let width_default = 0
let precision_default = 2
let base_default = Dec

# Uns.hmi ----------------------------------------------------------------------
type uns

val to_string: ?pad:codepoint -> ?just:Fmt.just -> ?sign:Fmt.sign -> ?alt:bool
  -> ?zpad:bool -> ?width:uns -> ?base:Fmt.base
  -> uns -> string
val fmt >e: ?pad:codepoint -> ?just:Fmt.just -> ?sign:Fmt.sign -> ?alt:bool
  -> ?zpad:bool -> ?width:uns -> ?base:Fmt.base
  -> uns -> Fmt.Formatter e >e-> Fmt.Formatter e

# String.hmi -------------------------------------------------------------------
type string

val to_string: ?pad:codepoint -> ?just:Fmt.just -> ?alt:bool -> ?width:uns
  -> string -> string
val fmt >e: ?pad:codepoint -> ?just:Fmt.just -> ?alt:bool -> ?width:uns
  -> string -> Fmt.Formatter e >e-> Fmt.Formatter e

val Fmt:
    type Formatter: Formatter =
      Fmt.Formatter {|} with type t := t

    val empty: Formatter
    val finish: Formatter -> string

    val to_string >e: (Fmt.Formatter e -> Fmt.Formatter e) -> string

# String.hm --------------------------------------------------------------------
let to_string ?(pad=Fmt.pad_default) ?(just=Fmt.just_default)
  ?(alt:Fmt.alt_default) ?(width=Fmt.width_default) s formatter =
    ...

let fmt ?pad ?just ?alt ?width s (formatter >e:Fmt.Formatter e) =
    formatter |> Fmt.Formatter.fmt (to_string ?pad ?just ?alt ?width s)

let Fmt = {|
    type t = string list

    let empty = []

    let finish t =
        String.concat_rev t

    let to_string fmt_partial =
        empty |> fmt_partial |> finish
  |}

# File.hmi ---------------------------------------------------------------------
type file

val Fmt:
    type Formatter: Formatter =
      Fmt.Formatter os with type t := t

    val makeFormatter: file -> Formatter
    val flush: Formatter >os-> Formatter
```

# Formatted strings

Assume the following lead-in code for all the subsequent formatted string syntax examples.

```hemlock
let print s =
    File.write_hlt (Bytes.of_string_slice (String.C.Slice.of_string s)) s File.stdout
let name = "Bob"
let age = 42
let children = ["Alice"; "Charlie"; "Megan"]
let children_fmt children (formatter >e:Fmt.Formatter e) =
    let rec fn children formatter =
        match children with
        | [] -> ()
        | child_m :: child_n :: [] ->
            formatter
              |> String.fmt child_m
              |> String.fmt " and "
              |> String.fmt child_n
        | child :: children ->
            formatter
              |> String.fmt child
              |> String.fmt ", "
              |> fn children
    fn children formatter
```

Here's how the formatter API could be used to produce a string:

```hemlock
print (String.Fmt.empty
  |> String.fmt "Hello, "
  |> String.fmt name
  |> String.fmt ", your advanced (binary) age of "
  |> Uns.fmt ~alt:true ~base:Bin age
  |> String.fmt " is excessive.\n\
Next year you will be "
  |> Uns.fmt (succ age)
  |> String.fmt " years old.
But wow, your progeny - "
  |> children_fmt children
  |> String.fmt " - are cute."
  |> String.Fmt.finish
  )

# Prints:
# `|Hello, Bob, your advanced (binary) age of 0b11010 is excessive.
#  |Next year you will be 43 years old.
#  |But wow, your progeny - Alice, Charlie and Megan - are cute.
# `
```

## Proposed syntax

OCaml's format specifiers have the unfortunate quality that they look like strings, but they are augmented with additional type information so that calls to `printf`-like functions can be statically typed. This requires rather complex types in the `printf` machinery, but the bigger problem is that the dual nature of strings becomes apparent if the application tries to use a format string reference rather than a string literal as the format specifier to `printf`. The syntax proposals that follow intentionally embed parameters directly into the string literals, so that it is syntactically impossible to disjoin a format specifier from its parameters.

```hemlock
print "Hello, %(name), your advanced (binary) age of %#b(age) is excessive.\n\
Next year you will be %(succ age) years old.
But wow, your progeny - %t(children_fmt children) - are cute."
```
This reads well, but it's quite tricky to parse. The challenge arises because arbitrary code in `%(...)` can contain nested parens, strings, etc., and we cannot in general recognize the matching `)` without parsing the code. That requires a nested expression parser context distinct from the outer parser, and parser nesting can recurse arbitrarily deep in the general case. Although this is a solvable parsing problem, it demands sophistication of external tooling, e.g. syntax highlighting in text editors, that is too onerous to seriously contemplate.

Consider what can be done just with the infix `^` string concatenation operator:

```hemlock
print ("Hello, "^name^", your advanced (binary) age of "
  ^(Uns.to_string ~alt:true ~base:Bin age)) ^" is excessive.\n\
Next year you will be "^(Uns.to_string (succ age))^" years old.
But wow, your progeny - "^(String.Fmt.to_string (children_fmt children))
  ^" - are cute.")
```

The code is quite succinct for `name`, but it gets pretty bulky otherwise. Still, this isn't a ridiculous option. Now consider the following syntax:

```hemlock
print "Hello, %(^name^), your advanced (binary) age of %#b(^age^) is excessive.\n\
Next year you will be %(^succ age^) years old.
But wow, your progeny - %t(^children_fmt children^) - are cute."
```

This syntax can be parsed without resorting to nested parsers, because the `%(^...^)` delimiters are distinct relative to every other syntactic construct in Hemlock. Furthermore the syntax evokes a synergy between formatted string syntax and string concatenation. In fact, the following mix is valid (note how `name` is concatenated into the result):

```hemlock
print "Hello, "^name^", your advanced (binary) age of %#b(^age^) is excessive.\n\
Next year you will be %(^succ age^) years old.
But wow, your progeny - %t(^children_fmt children^) - are cute."
```

## Semantics

The formatted string syntax has some significant restrictions with regard to how it formats values:

- Only the atomic builtin types have implicit formatting support:
  + `unit`
  + `bool`
  + Numeric types (`uns`, `int`, `u8`, `nat`, `real`, etc.)
  + `codepoint`
  + `string`
  These types have non-extensible support for common formmatting options (e.g. padding and numerical base), enumerated below.
- Polymorphic collection types (`List`, `Map`, etc.) provide parametrized `fmt` functions which can be composed and used via the `%t` specifier.
- Application types can provide `fmt` functions and use them with the `%a`/`%t` specifiers.

## Format specifiers

A format specifier can be embedded as `%<specifier>(^...^)`. Some specifier flags apply only to a subset of types, e.g. zero padding (`<zpad>`) only applies to numeric types.

```
%['<pad>'][<just>][<sign>][<alt>][<zpad>][<width>][.<precision>][<type-specific fmt>](^...^)
```

- `'<pad>'`: Pad with specified codepoint (default: `' '`)
- `<just>'`: Justify
  + `[`: Left
  + `][`: Center
  + `]`: Right (default)
- `<sign>`: Sign (default is to print sign only when negative)
  + `+`: Explicit sign, even when positive
  + `_`: Space in place of sign if sign is non-negative
- `<alt>`: `#` enables alternate formatting
  + Unit: `()`
  + Numeric types: prefix with base, separate digit groups with `_`
    * Binary: `0b` prefix, groups of 8
    * Octal: `0o` prefix, no groups
    * Decimal: No prefix, groups of 3
    * Hexadecimal: `0x` prefix, groups of 4
  + Codepoint: `'c'`
  + String: ``` ``some string`` ```
- `<zpad>`: Pad numeric type with leading zeros. This is distinct from `<pad>` because zero padding goes between the sign/prefix and the digits.
- `<width>`: Width
  + `42`: Fixed width in codepoints
  + `*`: Parametric width in codepoints, e.g. `%*(^width^)(^some_val^)`
- `.<precision>`: Number of digits past the point
  + `.2`: Fixed precision
  + `.*`: Parametric precision, e.g. `%*(^width^).*(^precision^)(^r^)`
- `<type-specific>`: Type-specific format
  + `b`: Binary (integer/real)
  + `o`: Octal (integer/real)
  + `d`: Decimal (integer/real) (default base)
  + `x`: Hexadecimal (integer/real)
  + `e`: Real in decimal exponential notation
  + `p`: Real in binary floating point notation (`bp`, `op`, `xp`*)
  + `g`: Real in (non-)exponential notation, whichever is more compact (`bg`, `og`, `dg`, `xg`).
  + `a`: Type-specific `fmt`, e.g. `%a(^(Option.fmt (Uns.fmt ~alt:true))^)(^maybe_u^)`
  + `t`: Type-generic formatter, e.g. `%t(^(Option.fmt (Uns.fmt ~alt:true)) maybe_u^)`

Note that some flags like `<zpad>` cannot be used universally, e.g. as `%0(^"hello"^)`, and such malformed specifiers will cause compilation failure. The `%a` case is unusual in that the `fmt` function being called may support any combination of flags, but just like for the builtin types, passing unsupported flags will trigger compilation failure. The `%t` case is unusual in that *no* flags are supported.

Examples:

```hemlock
"%(^^)%(^()^)" # Unit values, formatted as "".
"%#(^^)" # Unit value, formatted as "()".
"%(^name^)"
"%#x(^age^)"
"%(^succ age^)"
"%][50a(^(List.fmt String.fmt)^)(^children^)"
"%t(^(List.fmt String.fmt) children^)"
"%016#x(^some_uns^)"
"%'␠'][*(^page_width^)(^title^)\n"
"%'␠'][(^title^)\n"
"(%'*'98(^""^))" # Length-100 "(**...**)" string.
"(%'*'98(^^))" # Length-100 "(**...**)" string.
"%#x(^x^) > %#x(^y^) -> %(^x > y^)"
"%#b(^x^) %#o(^x^) %#(^x^) %#x(^x^)"
"\%" # Need to escape; could use "%%` instead, but "\%" feels more correct.
```

**Comment by jasone** [comment]
`Real` formatting is a deep topic that has consumed quite a bit of time over the past several days. I have normalized formatting implemented now, though the decimal output does not promise the same precision/stability as does `strfromd(3)`. I don't think this is a huge deal, since we provide bit-accurate base-2/8/16 formatting. That said, someday we may want to precisely match the formatting implemented in other languages. Doing so will require multi-word integer math, for which the `nat` type will be ideal (once it's integrated).

For future reference, the following documents how to convert to/from decimal format:

    Correctly Rounded Binary-Decimal and Decimal-Binary Conversions
    David M. Gay
    AT&T Bell Labs
    Numerical Analysis Manuscript 90-10
    November 30, 1990

**Comment by jasone** [comment]
`Bytes.Slice.fmt` would be trivial to implement were `Array.Slice` available. Rather than hacking a solution, I'm going to tackle #90 as a prerequisite.

**Comment by jasone** [comment]
Yesterday I ran into a problem with the planned interface for `Formattable`. The original idea was to just have each type supply a `fmt` function and integrate that into `Formatttable` but that has two problems:

- The optional parameters for e.g. `Uns.fmt` make the functions incompatible with the interface.
- Some functions, in particular `String.pp`, have non-default parametrizations with respect to `fmt`.

My initial plan was to add an `xfmt` for types which support optional formatting parameters, but the `String.fmt` vs `String.xfmt` situation violates POLA. After further thought I decided to have a `pp` function and a `fmt` function, where `pp` serves the "pretty printing" role for `Identifiable` types. That's going to require an intermediate renaming step to move the `Format`-based `pp` functions to `xpp` so that there is no name collision. On the bright side, it will be trivial to spot remaining uses of the old pretty printing code via `git grep xpp`.

**Comment by jasone** [comment]
It became clear while using `ZInt.fmt` to format the exponent for `Realer.pp`, pretty-printing the suffix needs to be controlled via a separate parameter. I added the `pretty` parameter, refactored the numeric types to use it, and enhanced `String.fmt` to additionally support formatting using the raw string syntax.

**Comment by jasone** [comment]
I ended up enhancing the compact-notation `real` formatting after dealing with some poor API ergonomics late in the test refactoring. This issue is now fully implemented aside from the scanner enhancements.

**Comment by jasone** [comment]
While reviewing the diffs I determined that the way I was using `precision` and `notation` to control whether trailing zeros are omitted was overly complex, and that the right solution is to add `pmode` (precision mode) which is either `Limited` or `Fixed`. I got that implemented, but during test refactoring I discovered that there are some previously unnoticed bugs in `RadixPoint` formatting for at least `Hex`. It's probably going to be a couple days until I can get that fixed.
